### ISSUE TITLE
ScrollReveal not working on iOS Chrome

### ISSUE THREAD

**Issue by jmalatia** [jlmakes/scrollreveal.js#196]
Sample site https://scrollrevealjs.org/ not working on iOS Chrome. iPhone or iPad iOS 9+.

**Comment by jmalatia** [comment]
Sample site https://scrollrevealjs.org/ not working on iOS Chrome. iPhone or iPad iOS 9+.

**Comment by jlmakes** [comment]
It seems that `el.getClientBoundingRect()` is unsuitable (re: #193) and I’m suspicious this issue may be related. I will be debugging Chrome on iOS, hopefully to roll out with the fix for #193 on Monday.

Thanks @jmalatia :bow:

**Comment by jlmakes** [comment]
I’ve confirmed it is working.

**Note**: Chrome on iOS however, does still pause animations during scroll. This behavior can be seen on many mobile browsers, as scroll event handling is postponed until the client has stopped scrolling. Mobile Safari only recently changed to support event handling during scroll as of iOS 8.

**Comment by jmalatia** [comment]
I just tried it again after updating to v3.0.3 and it is still not working for me on multiple iOS devices using Chrome (current updated versions of the app and iOS).  Not working both in my own project and on the sample site https://scrollrevealjs.org/

**Comment by jlmakes** [comment]
Hmm, @jmalatia it's not working for me either now… bah, okay I will look into this again. (I need a good way to debug mobile browsers!)

**Comment by BlackSnail** [comment]
how about compatibility of V2.x in mobile browsers?

**Comment by jlmakes** [comment]
@BlackSnail Version 2 appears to work fine in Chrome: https://scrollrevealjs.org/legacy

Which is actually confusing, since the core logic is strikingly similar. I just need a remote inspector to look through the ScrollReveal instance and any errors to figure out why Version 3 isn’t working properly in iOS Chrome.

**Comment by jlmakes** [comment]
@jmalatia Sometimes when it loads, it works… but most times it seems not to. This is such a confusing bug! I’m trying to use _weinre_ to debug it remotely, but I haven’t been able to connect to phone to my local network… I think I need to open port 8080?

@AyanGhatak If you could help out with this, it would be awesome!

**Comment by jlmakes** [comment]
@jmalatia @BlackSnail It seems the event throttling flag being outside `requestAnimationFrame` was causing this problem :weary: 

```diff
ScrollReveal.prototype.handler = function( event ) {
  if ( !sr.blocked ) {
-    sr.blocked = true;
-    _requestAnimationFrame( sr.animate );
+    _requestAnimationFrame(function(){
+      sr.blocked = true;
+      sr.animate();
+    });
  }
};
```

Had to debug the old fashioned way, but

**Comment by AyanGhatak** [comment]
Hey @jlmakes Things still don't work the similar manner in ios chrome and safari even after making the changes you suggested in [4b10dea](https://github.com/jlmakes/scrollreveal.js/commit/4b10deaa70a3e1e9ea819e4815705c3a350c644c). @jmalatia @BlackSnail Correct me if I am wrong!

Let me rephrase the exact issue we are facing :-
Things don't animate until the scrolling part is done in ios chrome. Whereas, in safari, we find animations occurring even while the scrolling is going.

The reason behind this :
In ios chrome, the scroll event is fired at the end of the scrolling. So unless the full scrolling is completed, the callback for the event(here, handler) is not invoked. Since the release of iOS 8 mobile Safari fixed the [scroll event issue](http://developer.telerik.com/featured/scroll-event-change-ios-8-big-deal/), but Chrome for iOS and Android still only fires the scroll event at the moment the drag stops. [Another Reference](https://github.com/luis-almeida/unveil/issues/90)

Parital workaround:
What I saw instead of scroll event, if we used touchmove in mobile devices and it was good. the handler callback was getting called. But....

there is again another bug for ios chrome: requestAnimationFrame(callback) : The callback is also called at a delayed time. In safari, the callback is invoked immediately. That's not the case with chrome. [Pollyfill](http://www.paulirish.com/2011/requestanimationframe-for-smart-animating/) didnot help much yet.

The workaround are the temporary findings and needs thorough validations.

**Comment by jlmakes** [comment]
#### The bug this issue originally referenced was in reference to reveals not occurring at all within iOS chrome after the initial load. 

:sparkles: :ambulance: _This has been fixed!_ :sparkles:

What you’re talking about @AyanGhatak is a long standing behavior of mobile browsers and scroll event handling. As I understand, iOS 8 was the *first* browser that evolved to trigger scroll events _during_ scrolling.

**The behavior you describe**—_where scroll events don’t fire until scrolling is complete_—**is not a bug with ScrollReveal, but a limitation of some mobile browsers.**

For example, this was posted in 2012: [On-Scroll Event Issues on Mobile Browsers](https://www.tjvantoll.com/2012/08/19/onscroll-event-issues-on-mobile-browsers/)

_Relevant Excerpt:_
>**Workaround Attempt 2 - Use Touch Events**
One consistent issue with this approach is that the ontouchmove event is only fired when the user’s finger remains on the screen. Meaning, if the user gives any momentum to the scroll, ontouchmove events will not be fired while the window is scrolling and their finger is not on the screen.

If you have a strategy for overcoming this that performs well without adding too much code to the source, I would be happy to explore the code!

**Comment by walkandre** [comment]
I hate to do this but I'm still seeing the problem with version v3.0.8 on ios 9.2 / chrome.

**Comment by jlmakes** [comment]
@walkandre there was a bug with checking support. Please update to `v3.0.9`.

**Comment by jmalatia** [comment]
I should have responded before, my bad.  @jlmakes I updated my project after you released v3.0.6  and tested with the revised version and it worked great!  Thanks for the speedy fix and keeping us updated during the process... much appreciated!

@walkandre, I just updated my project to v3.0.9 and tested in ios chrome and it is working correctly...  Not sure about v3.0.8.  One thing I have noticed when testing on ios (either safari or chrome) is that when things start to act weird (especially with pages that have animations or transitions that I have refreshed the page a lot on), I close all open web browser windows and then go back to the main ios home screen and dbl click the home button and throw up safari and/or chrome to release them from memory and then reopen the app and try again --- then usually I get what is expected and the bad/weird behavior is gone.

**Comment by jlmakes** [comment]
My pleasure @jmalatia, thanks for helping out! :bow:

**Comment by walkandre** [comment]
@jlmakes thanks for the awesome library! btw you should take a look here: https://developers.google.com/web/tools/chrome-devtools/debug/remote-debugging/remote-debugging?hl=en the link explains how to remote debug ios and android devices. it's pretty handy.

**Comment by romainpoirier** [comment]
Probably a related issue, but after testing the [demo website](https://scrollrevealjs.org) using [BrowserStack](https://www.browserstack.com/), I can confirm it doesn't work under Safari 10.1 (included, and older versions). This Safari version is still widely used on some older Macs.
